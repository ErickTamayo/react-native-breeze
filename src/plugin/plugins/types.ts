import { ImageStyle, StyleProp, TextStyle, ViewStyle } from "react-native";
import Config from "../utils/config";

type PatterCallableParams = {
  keys: typeof Config.keys;
};

export type PatternCallable = (params: PatterCallableParams) => RegExp;

export type PluginPattern = RegExp | PatternCallable;

export type PluginFunctionParams<T> = {
  input: string;
  groups: T;
  theme: typeof Config.theme;
  color: typeof Config.color;
};

export type PluginFunctionReturnType =
  | StyleProp<ViewStyle>
  | StyleProp<ImageStyle>
  | StyleProp<TextStyle>;

export type PluginFunction<T = {}> = (
  params: PluginFunctionParams<T>
) => PluginFunctionReturnType;
