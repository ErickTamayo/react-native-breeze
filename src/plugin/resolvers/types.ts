export type ResolverFunctionParams = {
  input: string;
  groups: { [group: string]: string };
  theme: (key: string) => any;
};

export type ResolverFunction = (params: ResolverFunctionParams) => object;

export type Resolver = {
  pattern: RegExp;
  fn: ResolverFunction;
};
