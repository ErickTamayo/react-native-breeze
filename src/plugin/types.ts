export type Configuration = {
  theme: {
    [key: string]: any;
  };
  variants: {
    [key: string]: any;
  };
  corePlugins: {
    [key: string]: boolean;
  };
  plugins: string[];
};
