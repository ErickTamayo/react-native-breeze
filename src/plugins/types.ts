import { ImageStyle, StyleProp, TextStyle, ViewStyle } from "react-native";
import { keys, colorKeys, theme, color } from "../config";

type PatterCallableParams = {
  keys: typeof keys;
  colorKeys: typeof colorKeys;
};

export type PatternCallable = (params: PatterCallableParams) => RegExp;

export type PluginPattern = RegExp | PatternCallable;

export type PluginFunctionParams<T> = {
  input: string;
  groups: T;
  theme: typeof theme;
  color: typeof color;
};

export type PluginFunctionReturnType =
  | StyleProp<ViewStyle>
  | StyleProp<ImageStyle>
  | StyleProp<TextStyle>;

export type PluginFunction<T = {}> = (
  params: PluginFunctionParams<T>
) => PluginFunctionReturnType;
