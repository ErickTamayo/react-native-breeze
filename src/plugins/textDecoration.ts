import { PluginFunction } from "./types";

// prettier-ignore
export type PluginGroups = {
  decoration: "no-underline" | "underline" | "line-through" | "underline-double" | "line-through-double" | "underline-dotted" | "line-through-dotted" | "underline-dashed" | "line-through-dashed";
};

export const pattern = /^(?<decoration>no-underline|underline|line-through|underline-double|line-through-double|underline-dotted|line-through-dotted|underline-dashed|line-through-dashed)$/;

export const plugin: PluginFunction<PluginGroups> = ({ groups }) => {
  const { decoration } = groups;

  switch (decoration) {
    case "no-underline":
      return { textDecorationLine: "none" };
    case "underline":
      return { textDecorationLine: "underline", textDecorationStyle: "solid" };
    case "underline-double":
      return { textDecorationLine: "underline", textDecorationStyle: "double" };
    case "underline-dotted":
      return { textDecorationLine: "underline", textDecorationStyle: "dotted" };
    case "underline-dashed":
      return { textDecorationLine: "underline", textDecorationStyle: "dashed" };
    case "line-through":
      return {
        textDecorationLine: "line-through",
        textDecorationStyle: "solid",
      };
    case "line-through-double":
      return {
        textDecorationLine: "line-through",
        textDecorationStyle: "double",
      };
    case "line-through-dotted":
      return {
        textDecorationLine: "line-through",
        textDecorationStyle: "dotted",
      };
    case "line-through-dashed":
      return {
        textDecorationLine: "line-through",
        textDecorationStyle: "dashed",
      };
  }
};
