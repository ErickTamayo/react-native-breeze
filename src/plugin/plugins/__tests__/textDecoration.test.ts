import { pattern, plugin, PluginGroups } from "../textDecoration";
import { PluginFunctionReturnType } from "../types";

describe("textDecoration", () => {
  it("Should match the correct pattern", () => {
    expect(pattern.exec("no-underline")).toBeTruthy();
    expect(pattern.exec("underline")).toBeTruthy();
    expect(pattern.exec("line-through")).toBeTruthy();
    expect(pattern.exec("underline-double")).toBeTruthy();
    expect(pattern.exec("line-through-double")).toBeTruthy();
    expect(pattern.exec("underline-dotted")).toBeTruthy();
    expect(pattern.exec("line-through-dotted")).toBeTruthy();
    expect(pattern.exec("underline-dashed")).toBeTruthy();
    expect(pattern.exec("line-through-dashed")).toBeTruthy();

    expect(pattern.exec("unknown")).toBeFalsy();
  });

  // prettier-ignore
  it.each<[input: string, expected: PluginFunctionReturnType]>([
    ["no-underline", { textDecorationLine: "none" }],
    ["underline", { textDecorationLine: "underline", textDecorationStyle: "solid" }],
    ["underline-double", { textDecorationLine: "underline", textDecorationStyle: "double" }],
    ["underline-dotted", { textDecorationLine: "underline", textDecorationStyle: "dotted" }],
    ["underline-dashed", { textDecorationLine: "underline", textDecorationStyle: "dashed" }],
    ["line-through", { textDecorationLine: "line-through", textDecorationStyle: "solid" }],
    ["line-through-double", { textDecorationLine: "line-through", textDecorationStyle: "double" }],
    ["line-through-dotted", { textDecorationLine: "line-through", textDecorationStyle: "dotted" }],
    ["line-through-dashed", { textDecorationLine: "line-through", textDecorationStyle: "dashed" }]
  ])("Should parse %s style correctly", (input, expected) => {
    const theme = jest.fn();
    const groups = (pattern as RegExp).exec(input)!.groups! as PluginGroups;
    const result = plugin({ input, groups, theme });

    expect(result).toEqual(expected);
  });
});
