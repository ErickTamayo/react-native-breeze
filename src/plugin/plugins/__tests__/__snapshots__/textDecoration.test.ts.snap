// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`textDecoration Should parse (line-through) correctly 1`] = `
Object {
  "textDecorationLine": "line-through",
  "textDecorationStyle": "solid",
}
`;

exports[`textDecoration Should parse (line-through-dashed) correctly 1`] = `
Object {
  "textDecorationLine": "line-through",
  "textDecorationStyle": "dashed",
}
`;

exports[`textDecoration Should parse (line-through-dotted) correctly 1`] = `
Object {
  "textDecorationLine": "line-through",
  "textDecorationStyle": "dotted",
}
`;

exports[`textDecoration Should parse (line-through-double) correctly 1`] = `
Object {
  "textDecorationLine": "line-through",
  "textDecorationStyle": "double",
}
`;

exports[`textDecoration Should parse (no-underline) correctly 1`] = `
Object {
  "textDecorationLine": "none",
}
`;

exports[`textDecoration Should parse (underline) correctly 1`] = `
Object {
  "textDecorationLine": "underline",
  "textDecorationStyle": "solid",
}
`;

exports[`textDecoration Should parse (underline-dashed) correctly 1`] = `
Object {
  "textDecorationLine": "underline",
  "textDecorationStyle": "dashed",
}
`;

exports[`textDecoration Should parse (underline-dotted) correctly 1`] = `
Object {
  "textDecorationLine": "underline",
  "textDecorationStyle": "dotted",
}
`;

exports[`textDecoration Should parse (underline-double) correctly 1`] = `
Object {
  "textDecorationLine": "underline",
  "textDecorationStyle": "double",
}
`;
